discovery_engine/lib/src/api/events/client_events.dart

```diff
@@ -95,6 +95,21 @@ class ClientEvent with _$ClientEvent {
     Set<DocumentId> documentIds,
   ) = FeedDocumentsClosed;
 
+  /// Event created when a source is added to the list of excluded sources.
+  @Implements<FeedClientEvent>()
+  const factory ClientEvent.excludedSourceAdded(Uri source) =
```

Comment:
As specified in the story the only values that are allowed here are value present in the field `source_domain`. That is not a valid `Uri` so we cannot use this type here.

Reply:
I've changed that to `String`

Reply:
Shouldn't it also be turned into a string in the other places, mainly the repository?

And if we treat it as an opaque value returned by news catcher then shouldn't
we also not sanitize it, because it's already in exact the format Newscatcher expects?

Reply:
this is only our assumption that the value sent by the app will be a `source_domain` from news catcher, so I think it's better to sanitize it and/or check if it's vaild. We could also change the value in the events to `Document` or `DocumentId` so then we know that we can get a proper `source_domain` and not a random `String`.

Reply:
The thing is a domain is not a `Uri`.

The only reason this works is because dart `Uri`s are not necessary Uris :sob: 

And as far as I understood `source_domain`'s must be _exactly_ the value from the `source_domain` field as we got from Newscatcher. Not the domain of the link.

"sanitizing" something which must not be changed isn't a good idea IMHO.

Also we can't use `DocumentId` because the excluded `source_domain` values might not be associated with any document we currently do have stored anywhere (neither in dart nor rust). 

So the only alternative would be to "newtype" `source_domain`.  Maybe in a follow up PR. But I'm not sure it's necessary.

Reply:
One simple way to avoid wrong data would be to check when we receive an `AddExcludedSource` that the value is a value of the field `source_domain` in the database. We can do a quick test with 10k documents and see how long it takes. If it is too much we can restrict that to only active documents (that kinda make sense anyway)

Reply:
@acrrd I've added that check, I think it should be fine, cause Hive loads all the documents into memory anyways. Let me know if the implementation is ok.

Reply:
> value is a value of the field source_domain in the database

But this will still couple the source to documents, which means when we e.g. prune the history a excluded source can sneak back in but IMHO should not. It's also the kind of error we likely won't catch with testing and will have problem tracing down when getting bug reports.

Reply:
> value of the field source_domain in the database

Through this still couples the source to a document, which IMHO we should avoid.

Reply:
> > value of the field source_domain in the database
> 
> Through this still couples the source to a document, which IMHO we should avoid.

ok, so what if we manage a map of sources and their exclusion state via `Map<String, bool>` (I can adapt existing repo to do that)? We can add new sources this map every time we get new documents using their `sourceDomain`s, we can include/exclude a source by toggling the flag, and we can check if a source is valid if it's present as a key in this map? This won't be coupled to the document, but it will still give us some assurance regarding validity of the sources send from the app.

```dart
final excludedSources = {
  'example.com': true, // excluded
  'bbc.co.uk': false, // included
};

if (excludedSources['theguardian.com'] == null) {
  // invalid source
}
```